File: Go/stack.go
package main
import (
    "fmt"
)

type Stack struct  {
    data []interface{}
}

func (s *Stack) size() int {
    return len(s.data)
}

func (s *Stack) push(item interface{}) {
    s.data = append(s.data, item)
}

func (s *Stack) peek() interface{} {
    return s.data[len(s.data) - 1]
}

func (s *Stack) pop() interface{} {
    idx := len(s.data) - 1
    res := s.data[idx]
    s.data = s.data[0:idx]
    return res
}

func (s *Stack) isEmpty() bool {
    return len(s.data) == 0
}

func checkParentheses (str string) bool {
    stack := new(Stack);
    for _, c := range str {
        if c == ')' || c == ']' || c == '}' || c == '>' {
            if stack.isEmpty() || stack.pop() != c {
                return false;
            }
        }
        if c == '(' { stack.push(')') }
        if c == '[' { stack.push(']') }
        if c == '{' { stack.push('}') }
        if c == '<' { stack.push('>') }
    }

    return stack.isEmpty()
}

func toBase(num int, base int) string {
    digits := "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    stack := new(Stack)
    for num > 0 {
        stack.push(digits[num % base])
        num /= base
    }

    res := ""
    for !stack.isEmpty() {
        res += fmt.Sprintf("%c", stack.pop())
    }

    return res
}

func toBinary(num int) string {
    return toBase(num, 2);
}

File: Python/stack.py
__author__ = 'Daniel'


class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def is_empty(self):
        return self.items == []

    def size(self):
        return len(self.items)

    def pop(self):
        return self.items.pop()

    def peek(self):
        return self.items[len(self.items) - 1]


def check_parentheses(inp):
    stack = Stack()
    for c in inp:
        if c == ')' or c == ']' or c == '}' or c == '>':
            if stack.is_empty() or stack.pop() != c:
                return False
        if c == '(':
            stack.push(')')
        if c == '[':
            stack.push(']')
        if c == '{':
            stack.push('}')
        if c == '<':
            stack.push('>')
    return stack.is_empty()


def to_base(num, base):
    digits = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    stack = Stack()
    while num > 0:
        stack.push(digits[num % base])
        num //= base

    res = ""
    while not stack.is_empty():
        res += stack.pop()

    return res


def to_binary(num):
    return to_base(num, 2)


def to_postfix(string):
    tokens = string.split()
    prec = {"+": 1, "-": 1, "*": 2, "/": 2, "(": 0, ")": 0}
    operators = Stack()

    res = []
    for token in tokens:
        if token == "(":
            operators.push("(")
        elif token == ")":
            op = operators.pop()
            while op != "(":
                res.append(op)
                op = operators.pop()
        elif token in "+-*/":
            while not operators.is_empty() and prec[token] <= prec[operators.peek()]:
                res.append(operators.pop())
            operators.push(token)
        else:
            res.append(token)

    while not operators.is_empty():
        res.append(operators.pop())

    return " ".join(res)


def eval_postfix(string):
    tokens = string.split()

    def eval_op(f):
        right = eval_stack.pop()
        left = eval_stack.pop()
        eval_stack.push(f(left, right))

    eval_stack = Stack()
    from operator import add, floordiv, mul, sub

    for token in tokens:
        if token == "+":
            eval_op(add)
        elif token == "-":
            eval_op(sub)
        elif token == "*":
            eval_op(mul)
        elif token == "/":
            eval_op(floordiv)
        else:
            eval_stack.push(int(token))

    return eval_stack.pop()
